homeassistant/components/jewish_calendar/service.py

```diff
@@ -0,0 +1,463 @@
+"""Services to return HDate instances & other information."""
+
+from __future__ import annotations
+
+import datetime as dt
+import logging
+from typing import cast
+
+from hdate import HDate, HebrewDate, Zmanim, htables
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_LANGUAGE, CONF_LOCATION
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.service import async_set_service_schema
+import homeassistant.util.dt as dt_util
+from homeassistant.util.json import JsonObjectType
+
+from .const import (
+    CONF_CANDLE_LIGHT_MINUTES,
+    CONF_DIASPORA,
+    CONF_HAVDALAH_OFFSET_MINUTES,
+    DOMAIN,
+)
+
+LOGGER = logging.getLogger(__name__)
+# Maps Hebrew & English month names to Months enum values.
+MONTH_INPUT_MAP = {
+    name: htables.Months(index + 1)
+    for index, m in enumerate(htables.MONTHS)
+    # Require users to specify which Adar they want in case of leap year.
+    if m.english != "Adar"
+    for name in (m.english, m.hebrew)
+}
+# Maps Hebrew & English holiday names to Holiday.
+HOLIDAY_INPUT_MAP = {
+    name: holiday
+    for holiday in htables.HOLIDAYS
+    for name in (
+        holiday.name,
+        holiday.description.english,
+        holiday.description.hebrew.long,
+        holiday.description.hebrew.short,
+    )
+    if name
+}
+
+OUTPUT_SCHEMA = {
+    vol.Required("include_hebrew_date_info", default=False): cv.boolean,
+    vol.Required("include_holiday_info", default=False): cv.boolean,
+    vol.Required("include_zmanim", default=False): cv.boolean,
+}
+
+OUTPUT_FIELDS = {
+    "include_hebrew_date_info": {"default": True, "selector": {"boolean": {}}},
+    "include_holiday_info": {"default": False, "selector": {"boolean": {}}},
+    "include_zmanim": {"default": False, "selector": {"boolean": {}}},
+}
+
+
+def async_setup_services(hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+    """Register the services."""
+    data = hass.data[DOMAIN][config_entry.entry_id]
+
+    location = data[CONF_LOCATION]
+    hebrew = data[CONF_LANGUAGE] == "hebrew"
+    candle_lighting_offset = data[CONF_CANDLE_LIGHT_MINUTES]
+    havdalah_offset = data[CONF_HAVDALAH_OFFSET_MINUTES]
+    diaspora = data[CONF_DIASPORA]
+
+    async def get_gregorian_date(call: ServiceCall) -> ServiceResponse:
+        """Service call that returns Hebrew date info for a Gregorian date."""
```

Comment:
Based on the comment the name sounds confusing. Shouldn't it be `get_hebrew_date()`?

Reply:
No; all of these get Hebrew dates (or other information, depending on what `include_*` fields you set).

This is why I originally named this `for_gregorian_date` (_et al_), but I was asked to change them.

Reply:
Ok, so not `get_hebrew_date`, but `get_gregorian_date` is really counterintuitive. Maybe `convert_gregorian_to_hebrew`?

Reply:
But then how would you name all of these services (which all return the same outputs given different inputs) consistently?

Also, if you pass `include_holiday_info: true` and the other `include_*` false, that name doesn't really apply.

Maybe name the services `get_for_*`?

Reply:
Not really sure here.
It's basically get Hebrew date info for Gregorian date. Can we have get_hebrew_date_info accept either Gregorian date, Gregorian date range, or Hebrew date as an input?

Reply:
That's actually how I first started building this, but I realized that it wasn't a good idea:

 - Users are likely to want multiple response types (eg, Hebrew date and holiday info), and we shouldn't require them to make multiple service calls.
 - The parameters would be extremely confusing, since all of the services would end up having 4 or 5 different _sets_ of parameters that are mutually exclusive, making it very hard to tell which parameters to set to get which results.
   - As opposed to my current approach, where the common parameters are 3 simple booleans and each service has a clear purpose with matching parameters.